programs/gateway/src/lib.rs

```diff
@@ -80,6 +130,93 @@ pub mod gateway {
         Ok(())
     }
 
+    /// Withdraws amount to destination program pda, and calls on_call on destination program
+    ///
+    /// # Arguments
+    /// * `ctx` - The instruction context.
+    /// * `amount` - The amount of SOL to withdraw.
+    /// * `sender` - Sender from ZEVM.
+    /// * `data` - Data to pass to destination program.
+    /// * `signature` - The TSS signature.
+    /// * `recovery_id` - The recovery ID for signature verification.
+    /// * `message_hash` - Message hash for signature verification.
+    /// * `nonce` - The current nonce value.
+    pub fn execute(
+        ctx: Context<Execute>,
+        amount: u64,
+        sender: [u8; 20],
+        data: Vec<u8>,
+        signature: [u8; 64],
+        recovery_id: u8,
+        message_hash: [u8; 32],
+        nonce: u64,
+    ) -> Result<()> {
+        let pda = &mut ctx.accounts.pda;
+
+        verify_and_update_nonce(pda, nonce)?;
+
+        let mut concatenated_buffer = Vec::new();
+        concatenated_buffer.extend_from_slice(b"ZETACHAIN");
+        concatenated_buffer.push(InstructionId::Execute as u8);
+        concatenated_buffer.extend_from_slice(&pda.chain_id.to_be_bytes());
+        concatenated_buffer.extend_from_slice(&nonce.to_be_bytes());
+        concatenated_buffer.extend_from_slice(&amount.to_be_bytes());
+        concatenated_buffer.extend_from_slice(&ctx.accounts.destination_program.key().to_bytes());
+        require!(
+            message_hash == hash(&concatenated_buffer[..]).to_bytes(),
+            Errors::MessageHashMismatch
+        );
+
+        msg!("Computed message hash: {:?}", message_hash);
+
+        recover_and_verify_eth_address(pda, &message_hash, recovery_id, &signature)?;
+
+        // NOTE: have to manually create Instruction, pack it and invoke since there is no crate for contract
+        // since any contract with on_call instruction can be called
+        let instruction_data = CallableInstruction::OnCall {
+            amount,
+            sender,
+            data,
+        }
+        .pack();
+
+        // account metas for remaining accounts
+        let account_metas: Vec<AccountMeta> = ctx
+            .remaining_accounts
+            .iter()
+            .map(|account_info| {
+                // Check if the account is writable, then create the appropriate AccountMeta
+                if account_info.is_writable {
+                    AccountMeta::new(*account_info.key, account_info.is_signer)
+                } else {
+                    AccountMeta::new_readonly(*account_info.key, account_info.is_signer)
+                }
+            })
+            .collect();
```

Comment:
similar for this , maybe this can be a separate function , it would just make things easier to test imo